rollup/types/encoding/codecv0/codecv0.go

```diff
@@ -0,0 +1,497 @@
+	var totalL1CommitGas uint64
+
+	// Add extra gas costs
+	totalL1CommitGas += 100000                 // constant to account for ops like _getAdmin, _implementation, _requireNotPaused, etc
+	totalL1CommitGas += 4 * 2100               // 4 one-time cold sload for commitBatch
+	totalL1CommitGas += 20000                  // 1 time sstore
+	totalL1CommitGas += 21000                  // base fee for tx
+	totalL1CommitGas += CalldataNonZeroByteGas // version in calldata
+
+	// adjusting gas:
+	// add 1 time cold sload (2100 gas) for L1MessageQueue
+	// add 1 time cold address access (2600 gas) for L1MessageQueue
+	// minus 1 time warm sload (100 gas) & 1 time warm address access (100 gas)
+	totalL1CommitGas += (2100 + 2600 - 100 - 100)
+	totalL1CommitGas += GetKeccak256Gas(89 + 32)           // parent batch header hash, length is estimated as 89 (constant part)+ 32 (1 skippedL1MessageBitmap)
+	totalL1CommitGas += CalldataNonZeroByteGas * (89 + 32) // parent batch header in calldata
+
+	// adjust batch data hash gas cost
+	totalL1CommitGas += GetKeccak256Gas(uint64(32 * len(b.Chunks)))
+
+	totalL1MessagePoppedBefore := b.TotalL1MessagePoppedBefore
+
+	for _, chunk := range b.Chunks {
+		chunkL1CommitGas, err := EstimateChunkL1CommitGas(chunk)
+		if err != nil {
+			return 0, err
+		}
+		totalL1CommitGas += chunkL1CommitGas
+
+		totalL1MessagePoppedInChunk := chunk.NumL1Messages(totalL1MessagePoppedBefore)
+		totalL1MessagePoppedBefore += totalL1MessagePoppedInChunk
+
+		totalL1CommitGas += CalldataNonZeroByteGas * (32 * (totalL1MessagePoppedInChunk + 255) / 256)
+		totalL1CommitGas += GetKeccak256Gas(89 + 32*(totalL1MessagePoppedInChunk+255)/256)
+
+		totalL1CommitCalldataSize, err := EstimateChunkL1CommitCalldataSize(chunk)
+		if err != nil {
+			return 0, err
+		}
+		totalL1CommitGas += GetMemoryExpansionCost(totalL1CommitCalldataSize)
+	}
+
+	return totalL1CommitGas, nil
+}
+
+// EstimateBatchL1CommitCalldataSize calculates the calldata size in l1 commit for this batch approximately.
+func EstimateBatchL1CommitCalldataSize(b *encoding.Batch) (uint64, error) {
+	var totalL1CommitCalldataSize uint64
+	for _, chunk := range b.Chunks {
+		chunkL1CommitCalldataSize, err := EstimateChunkL1CommitCalldataSize(chunk)
+		if err != nil {
+			return 0, err
+		}
+		totalL1CommitCalldataSize += chunkL1CommitCalldataSize
+	}
+	return totalL1CommitCalldataSize, nil
+}
+
+func getTxPayloadLength(txData *types.TransactionData) (uint64, error) {
+	rlpTxData, err := encoding.ConvertTxDataToRLPEncoding(txData)
+	if err != nil {
+		return 0, err
+	}
+	return uint64(len(rlpTxData)), nil
+}
+
+// DecodeBlockRangesFromEncodedChunks decodes the provided chunks into a list of block ranges. Each chunk
+// contains information about multiple blocks, which are decoded and their ranges (from the
+// start block to the end block) are returned.
+func DecodeBlockRangesFromEncodedChunks(chunks [][]byte) ([]*rawdb.ChunkBlockRange, error) {
```

Comment:
`rawdb.ChunkBlockRange` depends on l2geth, this will make it hard to move this into its own library.

How about moving this into `rollup_sync_service` and simply using it as `decodeBlockRangesFromEncodedChunks(codecVersion, chunks)`?

Reply:
makes sense. changed in [ef09494](https://github.com/scroll-tech/go-ethereum/pull/678/commits/ef09494ea638852fafbd1e247780ef7846e7219d).